### ISSUE TITLE
Consider whether to always use destructuring when extracting member expressions

### ISSUE THREAD

**Issue by sluukkonen** [nicoespeon/abracadabra#161]
Given a function for concatenating two arrays.

```javascript
export function concat(a1, a2) {
  const result = new Array(a1.length + a2.length)

  for (let i = 0; i < a1.length; i++) {
    result[i] = a1[i]
  }

  for (let i = 0; i < a2.length; i++) {
    result[i + a1.length] = a2[i]
  }

  return result
}
```

I'm repeating `a1.length` and `a2.length` a few times in the code, so I use "Extract variable" on both of them and select to replace all occurrences.

```javascript
export function concat(a1, a2) {
  const { length: a1Length } = a1
  const { length: a2Length } = a2
  const result = new Array(a1Length + a2Length)

  for (let i = 0; i < a1Length; i++) {
    result[i] = a1[i]
  }

  for (let i = 0; i < a2Length; i++) {
    result[i + a1Length] = a2[i]
  }

  return result
}
```

Since I'm extracting a property with the same name twice, I have to rename them. The result ends up being somewhat noisy, plus it is longer than the more "obvious" solution.

```javascript
export function concat(a1, a2) {
  const a1Length = a1.length
  const a2Length = a2.length
  const result = new Array(a1Length + a2Length)

  for (let i = 0; i < a1Length; i++) {
    result[i] = a1[i]
  }

  for (let i = 0; i < a2Length; i++) {
    result[i + a1Length] = a2[i]
  }

  return result
}
```

Perhaps it would be a good idea not to use destructuring if the extracted variable has a different name than the original property?

I suppose this might be difficult to do, since it looks like extraction is done before the new name has been decided. But YMMV.

## Additional information

- Version of the extension used: _v4.7.0_

**Comment by sluukkonen** [comment]
Alternatively, I'm thinking this could be solved with two refactorings that converted between destructured and non-destructured assignments, so one could first use extract and then immediately convert it to the non-destructured version.

**Comment by nicoespeon** [comment]
@sluukkonen very good point, I experienced the same pain recently so I can relate.

After a few thoughts, @fabien0102 and I are going to simply ask the user when the 2 options are available:
![image](https://user-images.githubusercontent.com/1094774/92766276-10103c00-f364-11ea-98d4-bdbbc5861e8a.png)

The modal won't trigger if the thing can't be destructured.

**Comment by sluukkonen** [comment]
A third idea I had (related to #162): do not use destructuring when extracting the first member from an object. Convert to the destructured form when extracting multiple members from an object.

I really don't know which solution one is my favourite.

**Comment by nicoespeon** [comment]
That's a good idea indeed. Hard to find the correct balance. I feel that if I'm not renaming, I want to destructure. So in the end, hard to anticipate and maybe in this case asking the user is the best thing to do 🤷‍♂️ 

The downside would be if it's too disruptive, but I feel it's ok if it triggers only when relevant.

As for #162 that's a very good suggestion indeed. Actually, I remember we discussed that in the past with @fabien0102 but for some reason I forgot to create a dedicated issue. So I'm glad you bring that back 👍

**Comment by nicoespeon** [comment]
FYI, this is now available with the latest version published today (v4.9.0).
